### DISCUSSION TITLE
Loss of 0 when combining Version and SubVersion

### DISCUSSION THREAD

**Discussion by duylong** [zwave-js/zwave-js#6687] [Request Support / Investigate Issue] [answered]
### Checklist

- [X] I have read and followed the above instructions
- [X] I have checked the [troubleshooting section](https://zwave-js.github.io/node-zwave-js/#/troubleshooting/index) and my problem is **not** described there.

- [X] I have read the [changelog](https://github.com/zwave-js/node-zwave-js/blob/master/CHANGELOG.md) and my problem was **not** mentioned there or the fix **did not work**.

### Describe the issue

Hi,

I recently contacted Aeotec support and found that in some cases the firmware numbers were different between Zwave JS UI and Aeotec.

They explained to me that some softwares drop the 0 when combining "Version" and "SubVersion" together.

For example: 
```
    V1.60 means
        Version 01
        SubVersion 60
    V1.06 means
        Version 01
        SubVersion 06
```
In my case, Zwave JS UI tells me 1.6 when the correct version would be 1.06, thus V1.6 = V1.06 ? Really ?

### Software versions

zwave-js-ui: 9.9.0.09b107a
zwave-js: 12.4.4

### Device information

Observed with Home Energy Meter - Gen5 ZW095

### Checklist

- [X] I [made sure](https://zwave-js.github.io/node-zwave-js/#/troubleshooting/index?id=rule-1-use-the-driver-log-luke) to provide a **driver log** on level **debug**.

- [ ] The log includes a re-interview of the problematic device _(if applicable)_.

- [ ] The log includes the problematic interaction with the device _(if applicable)_.

- [ ] I provided the node ID of the problematic device _(if applicable)_.

### Upload Logfile

Just an observation after contacting Aeotec support, I have no logs.

[a comment by an automated account is not shown]

**Comment by AlCalzone** [comment]
Versions in Z-Wave are just 2 (major, minor) or 3 (major, minor, patch) numbers from 0-255. There is no required formatting of the numbers, e.g. leading zeros, which lead to ambiguity - why isn't 1.06 not formatted as 1.006 considering the version could go up to 1.255?

Therefore we format versions the easiest way possible by concatenating the individual parts with dots inbetween, just like [SemVer](https://semver.org/#semantic-versioning-specification-semver). Meaning:
- Major 1, Minor 6 becomes `1.6`
- Major 1, Minor 60 becomes `1.60`
- Major 2, Minor 255, Patch 7 becomes `2.255.7`

**Comment by duylong** [comment]
I also have this problem with Shelly, for example, I have a Shelly Qubino Wave 1 module with version 11.03 on their site but in Zwave UI, it is displayed 11.3.0. This problem is disturbing.

**Reply by josh-m-sharpe** [reply]
makes perfect sense that you would be confused by this.  zwave changing things on us leads to confusion.

**Comment by josh-m-sharpe** [comment]
I get semver.  I <3 semver.  It's great.  It'd be *wonderful* if all products used it everywhere.

But they don't.

Aeotec delives a firmware expressly called named "1.02" and "6.81.06"

IMO, These are the only `strings` that should be shown in the zwave UI.  It's what the vendor called it, so it's what should be shown.
It's *super* confusing when I see a different version between zwave and the vendor and wonder what on earth is going on.

ZwaveJS Ui being a middleman in this situation and altering what was delivered by the vendor is an extremely confusing idea.

**Reply by josh-m-sharpe** [reply]
related: https://github.com/zwave-js/zwave-js-ui/issues/3678

**Reply by josh-m-sharpe** [reply]
I'd add that a vendor releasing a version with leading zeros, to some degree, probably implies that they expect that major/minor/patch to have that many significant digits. Maybe.

In which case we'd expect these mappings:
".06" => 6
".2" => 20

and they'd be an animal if they actually released ".2" in this scheme instead of ".20"

But again, see point above about how developers are human and dork up semver all the time.  All the more reason zwave shouldn't be a middleman and attempt to interpret.

**Reply by AlCalzone** [reply, marked as answer]
> Aeotec delives a firmware expressly called named "1.02"

**They** call it like this in **their** documentation. There's no way for Z-Wave JS to know they do, since all the information we get from the stick is a number `1` (major version) and a number `2` (minor version).

![grafik](https://github.com/zwave-js/node-zwave-js/assets/17641229/bf6d2087-6978-482f-8699-4309a17a9d26)

Similarly, this is the information we get for the firmware version of end devices:
![grafik](https://github.com/zwave-js/node-zwave-js/assets/17641229/f7cabb69-9974-47c4-8c4d-eb35c99ddcc0)

![grafik](https://github.com/zwave-js/node-zwave-js/assets/17641229/8f862e4f-fcda-45f8-a15f-6ba141923c31)
![grafik](https://github.com/zwave-js/node-zwave-js/assets/17641229/1f74ef51-34db-4064-bf64-70e287545e49)

And this is the SDK version (which Silabs finally started formatting like SemVer does in 7.x.y):
![grafik](https://github.com/zwave-js/node-zwave-js/assets/17641229/9315e966-a8b6-4e2b-b139-b491550be350)

> altering what was delivered by the vendor 

We alter nothing, we work with the information we've got. I'm not going to hardcode that the minor version should be zero-padded to length 2 for manufacturer A, to length 3 for manufacturer B, not at all for manufacturer C and that manufacturer D likes to convert their version numbers to roman numerals or whatever their marketing team thinks of.

---

To end this discussion:
![grafik](https://github.com/zwave-js/node-zwave-js/assets/17641229/a060f9fb-6698-4d38-bbf9-b3fd5f64bb82)
